### PR TITLE
Table Export JSON Array + URL Generate + Generate Header Empty CSV

### PR DESCRIPTION
**Pull request by moizarafat**
Co-authored-by: chandrakaran

## Description
Table Export
i) JSON Array Output Generation. 
ii) Download URL Generation logic.
iii) Generate Header when no CSV results found.

## Motivation and Context
Table Export functionality

## How Has This Been Tested?
Existing Test case pass. New Test cases added.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-standalone/src/main/java/com/yahoo/elide/standalone/config/ElideResourceConfig.java [hunk 1]
```diff
@@ -142,8 +142,9 @@ protected void configure() {
 
                     // TODO: If null, initialize with FileResultStorageEngine
                     ResultStorageEngine resultStorageEngine = asyncProperties.getResultStorageEngine();
+                    // TODO: Add logic to generate baseURLPresent Flag
```
Comment (moizarafat) [inline elide-standalone/src/main/java/com/yahoo/elide/standalone/config/ElideResourceConfig.java:145]: Will do separate PR for Standalone to set baseURLPresent flag and binding of Table Export Hooks.

File: elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java [hunk 2]
```diff
@@ -26,44 +26,39 @@
 
 /**
  * Class for Table Export functionality.
+ * Parses a TableExport request and returns an Observable of PersistentResource.
  */
 @Slf4j
 public class TableExporter {
 
     private Elide elide;
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java:34]: AsyncExecutorService seems to be a better place to keep the downloadURL instead of TableExporter. TableExporter will not have any use of downloadURL.

File: elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java [hunk 3]
```diff
@@ -26,44 +26,39 @@
...
         try (DataStoreTransaction tx = elide.getDataStore().beginTransaction()) {
             elide.getTransactionRegistry().addRunningTransaction(requestId, tx);
 
-            EntityProjection projection = null;
             RequestScope requestScope = null;
 
+            TableExportParser parser = getParser(query);
+
+            EntityProjection projection = parser.parse(query, apiVersion);
+
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java:60]: The TableExportParser hides away the differences for different protocols (JSON-API, GraphQL, etc).  However, we still are check the queryType below to build the request scope.

We might want a different abstraction here.  Maybe the parser should return a function that runs the query - rather than a projection - so we can move together all the query type specific logic.

File: elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java [hunk 4]
```diff
@@ -99,4 +94,16 @@ public TableExporter(Elide elide, String apiVersion, User user, GraphQLParser gr
 
         return results;
     }
+
+    public TableExportParser getParser(TableExport query) {
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/export/TableExporter.java:98]: Maybe instead of returning a parser, this should return a load function which returns an Observable of records.

File: elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java [hunk 5]
```diff
@@ -15,29 +15,69 @@
...
 
 /**
  * LifeCycle Hook for execution of TableExpoer.
  */
+@Slf4j
 public class TableExportHook extends AsyncAPIHook<TableExport> {
+    private boolean baseURLPresent = true;
 
-    public TableExportHook (AsyncExecutorService asyncExecutorService, Integer maxAsyncAfterSeconds) {
+    public TableExportHook (AsyncExecutorService asyncExecutorService, Integer maxAsyncAfterSeconds,
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java:32]: baseUrlPresent is an odd contract.  Also, it doesn't really belong in the hook.

baseUrl can be extracted from the requestScope.elideSettings.

Constructing the downloadURL should be a utility function that is shared with the export controller and this class.

File: elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java [hunk 6]
```diff
@@ -15,29 +15,69 @@
...
         AsyncExecutorService service = getAsyncExecutorService();
-        TableExporter exporter = new TableExporter(service.getElide(), requestScope.getApiVersion(),
-                requestScope.getUser());
-        TableExportThread queryWorker = new TableExportThread(query, service.getResultStorageEngine(), exporter);
+
+        // TODO: Make it a Bean
+        TableExporter exporter = new TableExporter(service.getElide());
+
+        String downloadURL = retrieveDownloadURL(requestScope.getBaseUrlEndPoint());
+        TableExportThread queryWorker = new TableExportThread(query, exporter, downloadURL,
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java:47]: The AsyncExecutorService should not have a function called isSkipCSVHeader.  That is specific configuration for TableExport.

File: elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java [hunk 7]
```diff
@@ -15,29 +15,69 @@
...
+                service.isSkipCSVHeader());
         executeHook(operation, phase, query, requestScope, queryWorker);
     }
 
     @Override
     public void validateOptions(AsyncAPI query, RequestScope requestScope) {
         super.validateOptions(query, requestScope);
     }
+
+    private String retrieveDownloadURL(String baseURL) {
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java:58]: Let's move this logic.

File: elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java [hunk 8]
```diff
@@ -15,29 +15,69 @@
...
+            downloadURL = getAsyncExecutorService().getDownloadURL();
+        } else {
+            downloadURL = getBasePath(baseURL) + getAsyncExecutorService().getDownloadURL() + "/";
+        }
+
+        return downloadURL;
+    }
+
+    private String getBasePath(String URL) {
+        URIBuilder uri;
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/hooks/TableExportHook.java:70]: Same comment - this doesn't belong in the hook.

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncExecutorService.java [hunk 9]
```diff
@@ -50,6 +50,9 @@
     private static AsyncExecutorService asyncExecutorService = null;
     private ResultStorageEngine resultStorageEngine;
     private ThreadLocal<AsyncAPIResultFuture> asyncResultFutureThreadLocal = new ThreadLocal<>();
+    private String downloadURL;
+    // TODO Update constructor to populate these.
+    private boolean skipCSVHeader;
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncExecutorService.java:55]: Neither of these new variables belong in the generic AsyncExecutorService.

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncExecutorService.java [hunk 10]
```diff
@@ -84,10 +88,10 @@ private AsyncExecutorService(Elide elide, Integer threadPoolSize, AsyncAPIDAO as
      * @param asyncAPIDao DAO Object
      */
     public static void init(Elide elide, Integer threadPoolSize, AsyncAPIDAO asyncAPIDao,
-            ResultStorageEngine resultStorageEngine) {
+            ResultStorageEngine resultStorageEngine, String downloadURL) {
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncExecutorService.java:91]: The async executor shouldn't need the downloadURL.

File: elide-async/src/main/java/com/yahoo/elide/async/service/thread/AsyncQueryThread.java [hunk 11]
```diff
@@ -62,10 +69,31 @@ public AsyncAPIResult call() throws URISyntaxException, NoHttpResponseException
...
+     * @param response is the ElideResponse
+     * @param queryType is the query type (GraphQL or JSON).
+     * @return Record Count
+     * @throws IOException Exception thrown by JsonPath
+     */
+    protected Integer calculateRecordCount(ElideResponse response, QueryType queryType)
+            throws IOException {
+        Integer count = null;
+        if (response.getResponseCode() == 200) {
+            if (queryType.equals(QueryType.GRAPHQL_V1_0)) {
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/thread/AsyncQueryThread.java:87]: The code has these query type checks sprinkled in many different places.  
This suggests to me that we ought to have an interface (and concrete functions) with a set of functions that we need each query type to perform.

File: elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java [hunk 12]
```diff
@@ -36,33 +41,68 @@
  */
 @Slf4j
 public class TableExportThread implements Callable<AsyncAPIResult> {
+    private static final String FORWARD_SLASH = "/";
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java:44]: Why do we have both TableExportThread and AsyncQuerythread?  Do we need both concepts?

File: elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/AsyncProperties.java [hunk 13]
```diff
@@ -54,4 +54,9 @@
      * Whether or not the async feature is enabled.
      */
     private boolean enabled = false;
+
+    /**
+     * Settings for the export controller.
+     */
+    private ControllerProperties export = new ControllerProperties();
```
Comment (aklish) [inline elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/AsyncProperties.java:61]: There shouldn't be a call to new here.

File: elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/ElideAsyncConfiguration.java [hunk 14]
```diff
@@ -50,8 +51,12 @@
     public AsyncExecutorService buildAsyncExecutorService(Elide elide, ElideConfigProperties settings,
             AsyncAPIDAO asyncQueryDao, EntityDictionary dictionary,
             @Autowired(required = false) ResultStorageEngine resultStorageEngine) {
+        boolean baseURLPresent = settings.getBaseUrl() != null && !settings.getBaseUrl().isEmpty();
+        String downloadURL = baseURLPresent ? settings.getBaseUrl() + settings.getAsync().getExport().getPath()
```
Comment (aklish) [inline elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/ElideAsyncConfiguration.java:55]: This doesn't really mirror how we initialized JSONApiLinks.

File: elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java [hunk 15]
```diff
@@ -36,33 +41,68 @@
...
-        //TODO Add URL generation logic and set
-        //queryResult.setUrl(new URL("url"));
+        queryResult.setUrl(generateDownloadURL());
         queryResult.setRecordCount(downloadRecordCount);
         return queryResult;
     }
 
+    private URL generateDownloadURL() {
+        log.debug("generateDownloadUrl");
+        String asyncQueryID = queryObj.getId();
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java:92]: I think there are three places currently where we attempt to generate the downloadURL.  We need to consolidate this to one.

File: elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java [hunk 16]
```diff
@@ -157,4 +210,33 @@ protected String convertToCSV(PersistentResource resource) {
...
         this.downloadRecordCount++;
     }
+
+    /**
+     * Generate CSV Header when Observable is Empty.
+     * @param parser TableExportParser object.
+     * @return returns Header string which is in CSV format.
+     */
+    protected Observable<String> generateCSVHeader(TableExportParser parser) {
+        EntityProjection projection = parser.parse(queryObj, apiVersion);
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java:220]: Is this the 2nd time we construct the entity projection?  Ideally we should do this once.

File: elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java [hunk 17]
```diff
@@ -78,25 +118,39 @@ protected TableExport storeResults(Observable<String> result) {
      * @return result as Observable of String
      */
     protected Observable<String> processObservablePersistentResource(Observable<PersistentResource> resources) {
-        Observable<String> results = Observable.just("No Records Generated");
+        Observable<String> results = Observable.empty();
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java:121]: We should generate the CSV header using the same code regardless of what the record count is.

**Review by aklish — CHANGES_REQUESTED** [review]
